### ISSUE TITLE
Exception when fetching metrics from carbon's memory cache

### ISSUE THREAD

**Issue by zdenekpizl** [graphite-project/graphite-web#1223]
Hallo,

 using master branch of graphite-web, carbon and whisper I am getting exceptions like:

`Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/graphite/render/datalib.py", line 136, in fetchData
    seriesList = _fetchData(pathExpr,startTime, endTime, requestContext, seriesList)
  File "/usr/lib/python2.6/site-packages/graphite/render/datalib.py", line 99, in _fetchData
    fetches = [(node, node.fetch(startTime, endTime)) for node in matching_nodes if node.is_leaf]
  File "/usr/lib/python2.6/site-packages/graphite/node.py", line 30, in fetch
    return self.reader.fetch(startTime, endTime)
  File "/usr/lib/python2.6/site-packages/graphite/readers.py", line 178, in fetch
    for (timestamp, value) in cached_datapoints:
TypeError: 'float' object is not iterable
Thu May 14 12:45:50 2015 :: Failed after 2 retry! See: 'float' object is not iterable
Traceback (most recent call last):
  File "/usr/lib/python2.6/site-packages/graphite/render/datalib.py", line 136, in fetchData
    seriesList = _fetchData(pathExpr,startTime, endTime, requestContext, seriesList)
  File "/usr/lib/python2.6/site-packages/graphite/render/datalib.py", line 99, in _fetchData
    fetches = [(node, node.fetch(startTime, endTime)) for node in matching_nodes if node.is_leaf]
  File "/usr/lib/python2.6/site-packages/graphite/node.py", line 30, in fetch
    return self.reader.fetch(startTime, endTime)
  File "/usr/lib/python2.6/site-packages/graphite/readers.py", line 178, in fetch
    for (timestamp, value) in cached_datapoints:
TypeError: 'float' object is not iterable`

From my point of view these error occur when graphite-web is asking carbon's memory cache to complete data written to disk with data not yet written to disk. It does not matter what metric is it apparently.

Anybody have met this faulty behaviour? Thank you, with regards .zp.

**Comment by tomasdubec** [comment]
This happens only for metrics starting with CARBON_METRIC_PREFIX, because they are handled by send_request_to_all instead of send_request and send_request_to_all is not updated for new interface between graphite-web and carbon (dict instead of list). I will create PR momentarily..

**Comment by tomasdubec** [comment]
https://github.com/graphite-project/graphite-web/pull/1224

**Comment by zdenekpizl** [comment]
That's great, could anybody from elders check this PR proposal and elaborate in it? Thank you, regards, .zp.

**Comment by ttarczynski** [comment]
I have the same issue with the latest versions of graphite-web, carbon and whisper from master:

```
Fri Aug 14 08:15:38 2015 :: Got an exception when fetching data! See: 'float' object is not iterable Will do it again! Run: 1 of 2
Traceback (most recent call last):
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 136, in fetchData
    seriesList = _fetchData(pathExpr,startTime, endTime, requestContext, seriesList)
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 99, in _fetchData
    fetches = [(node, node.fetch(startTime, endTime)) for node in matching_nodes if node.is_leaf]
  File "/opt/graphite/webapp/graphite/node.py", line 30, in fetch
    return self.reader.fetch(startTime, endTime)
  File "/opt/graphite/webapp/graphite/readers.py", line 177, in fetch
    for (timestamp, value) in cached_datapoints:
TypeError: 'float' object is not iterable
Fri Aug 14 08:15:38 2015 :: Failed after 2 retry! See: 'float' object is not iterable
Traceback (most recent call last):
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 136, in fetchData
    seriesList = _fetchData(pathExpr,startTime, endTime, requestContext, seriesList)
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 99, in _fetchData
    fetches = [(node, node.fetch(startTime, endTime)) for node in matching_nodes if node.is_leaf]
  File "/opt/graphite/webapp/graphite/node.py", line 30, in fetch
    return self.reader.fetch(startTime, endTime)
  File "/opt/graphite/webapp/graphite/readers.py", line 177, in fetch
    for (timestamp, value) in cached_datapoints:
TypeError: 'float' object is not iterable
Fri Aug 14 08:15:38 2015 :: Exception encountered in <GET http://graph1/render/?width=586&height=308&_salt=1439540137.906&target=carbon.agents.graph1-a.droppedCreates>
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/django/core/handlers/base.py", line 112, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/opt/graphite/webapp/graphite/render/views.py", line 112, in renderView
    seriesList = evaluateTarget(requestContext, target)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 8, in evaluateTarget
    result = evaluateTokens(requestContext, tokens)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 29, in evaluateTokens
    return evaluateTokens(requestContext, tokens.expression, replacements)
  File "/opt/graphite/webapp/graphite/render/evaluator.py", line 45, in evaluateTokens
    return fetchData(requestContext, expression)
  File "/opt/graphite/webapp/graphite/render/datalib.py", line 141, in fetchData
    raise Exception("Failed after %i retry! See: %s" % (settings.MAX_FETCH_RETRIES, e))
Exception: Failed after 2 retry! See: 'float' object is not iterable
```

And it also happens only for metrics within CARBON_METRIC_PREFIX.
I've tried applying #1224 patch and it has fixed the problem for me.

**Comment by biancalana** [comment]
@obfuscurity this issue is still happening with current master

**Comment by obfuscurity** [comment]
Should be fixed by #1224.
